File: api/csharp/EviFluor/lookup_table.cs
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: © 2025 HSE AG, <opensource@hseag.com>

using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text.Json.Nodes;

namespace Hse.EviFluor.Kits;

/// <summary>
/// Single lookup-table point mapping normalized signal to concentration.
/// </summary>
/// <remarks>
/// Initializes a lookup-table entry.
/// </remarks>
public sealed class LookupTableEntry(double concentration, double signal)
{

    /// <summary>
    /// Concentration represented by the lookup-table point.
    /// </summary>
    public double Concentration { get; } = concentration;

    /// <summary>
    /// Normalized signal represented by the lookup-table point.
    /// </summary>
    public double Signal { get; } = signal;

    internal JsonObject ToJsonObject()
    {
        return new JsonObject
        {
            ["concentration"] = Concentration,
            ["signal"] = Signal,
        };
    }

    internal static LookupTableEntry FromJson(JsonNode? node)
    {
        JsonObject obj = node as JsonObject ?? throw new InvalidOperationException("Lookup table entries must be objects");
        return new LookupTableEntry(
            concentration: obj["concentration"]?.GetValue<double>() ?? throw new InvalidOperationException("Lookup table entries must contain concentration"),
            signal: obj["signal"]?.GetValue<double>() ?? throw new InvalidOperationException("Lookup table entries must contain signal"));
    }
}

/// <summary>
/// Stores lookup-table points and converts between CSV and JSON representations.
/// </summary>
/// <remarks>
/// Initializes the lookup table with optional point entries.
/// </remarks>
public sealed class LookupTable(IEnumerable<LookupTableEntry>? entries = null)
{

    /// <summary>
    /// Sorted lookup-table entries. The smallest signal is always at the beginning.
    /// </summary>
    public IReadOnlyList<LookupTableEntry> Entries { get; } = (entries ?? [])
            .OrderBy(entry => entry.Signal)
            .ToList()
            .AsReadOnly();

    /// <summary>
    /// Returns the lookup table as a JSON-serializable array.
    /// </summary>
    public JsonArray ToJson()
    {
        JsonArray array = [];
        foreach (var entry in Entries)
        {
            array.Add(entry.ToJsonObject());
        }
        return array;
    }

    /// <summary>
    /// Writes the lookup table to a JSON file.
    /// </summary>
    public void SaveJson(string jsonPath)
    {
        var parentDirectory = Path.GetDirectoryName(jsonPath);
        if (!string.IsNullOrEmpty(parentDirectory))
        {
            Directory.CreateDirectory(parentDirectory);
        }

        File.WriteAllText(jsonPath, ToJson().ToJsonString(new System.Text.Json.JsonSerializerOptions { WriteIndented = true }));
    }

    /// <summary>
    /// Loads a lookup table from a CSV or JSON file based on its extension.
    /// </summary>
    public static LookupTable Load(string path)
    {
        string extension = Path.GetExtension(path).ToLowerInvariant();
        return extension switch
        {
            ".json" => LoadJson(path),
            ".csv" => FromCsv(path),
            _ => throw new InvalidOperationException($"Unsupported lookup table file extension: {(string.IsNullOrEmpty(extension) ? "<none>" : extension)}"),
        };
    }

    /// <summary>
    /// Loads a lookup table from a JSON file.
    /// </summary>
    public static LookupTable LoadJson(string jsonPath)
    {
        JsonNode? node = JsonNode.Parse(File.ReadAllText(jsonPath));
        return node == null ? throw new InvalidOperationException("Lookup table JSON is empty") : FromJson(node);
    }

    /// <summary>
    /// Creates a lookup table from JSON data.
    /// </summary>
    public static LookupTable FromJson(JsonNode node)
    {
        if (node is JsonObject wrapper)
        {
            node = wrapper["lookupTable"] ?? throw new InvalidOperationException("Lookup table JSON must contain lookupTable");
        }

        JsonArray array = node as JsonArray ?? throw new InvalidOperationException("Lookup table JSON must contain a list of entries");
        return new LookupTable(array.Select(LookupTableEntry.FromJson));
    }

    /// <summary>
    /// Builds a lookup table from an RFU CSV file.
    /// </summary>
    public static LookupTable FromCsv(string csvPath)
    {
        string[] lines = File.ReadAllLines(csvPath);
        if (lines.Length == 0)
        {
            throw new InvalidOperationException("CSV file is empty");
        }

        string[] header = lines[0].Split(';');
        int rfuIndex = Array.IndexOf(header, "RFU");
        int concentrationIndex = Array.IndexOf(header, "Concentration");

        if (rfuIndex < 0 || concentrationIndex < 0)
        {
            throw new InvalidOperationException("CSV file is missing columns: RFU, Concentration");
        }

        List<(double Concentration, double Rfu)> rows = [];
        for (int rowIndex = 1; rowIndex < lines.Length; rowIndex++)
        {
            if (string.IsNullOrWhiteSpace(lines[rowIndex]))
            {
                continue;
            }

            string[] row = lines[rowIndex].Split(';');
            string rfuText = GetColumnValue(row, rfuIndex).Trim();
            string concentrationText = GetColumnValue(row, concentrationIndex).Trim();

            if (rfuText.Length == 0 && concentrationText.Length == 0)
            {
                continue;
            }

            if (rfuText.Length == 0 || concentrationText.Length == 0)
            {
                throw new InvalidOperationException($"CSV row {rowIndex + 1} must contain RFU and Concentration");
            }

            rows.Add((
                Concentration: double.Parse(concentrationText, CultureInfo.InvariantCulture),
                Rfu: double.Parse(rfuText, CultureInfo.InvariantCulture)));
        }

        if (rows.Count < 2)
        {
            throw new InvalidOperationException("CSV file must contain at least two data rows to derive RFU normalization");
        }

        double rfuStdHigh = rows[0].Rfu;
        double rfuStdLow = rows[1].Rfu;
        if (rfuStdHigh == rfuStdLow)
        {
            throw new InvalidOperationException("RFU standard high and standard low must be different");
        }

        List<LookupTableEntry> entries = [];
        foreach (var row in rows)
        {
            entries.Add(new LookupTableEntry(
                concentration: row.Concentration,
                signal: (row.Rfu - rfuStdLow) / (rfuStdHigh - rfuStdLow)));
        }

        return new LookupTable(entries);
    }

    /// <summary>
    /// Creates a JSON lookup-table file from an RFU CSV file.
    /// </summary>
    public static LookupTable CreateJsonFromCsv(string csvPath, string jsonPath)
    {
        LookupTable lookupTable = FromCsv(csvPath);
        lookupTable.SaveJson(jsonPath);
        return lookupTable;
    }

    private static string GetColumnValue(string[] row, int index)
    {
        return index < row.Length ? row[index] : string.Empty;
    }
}

File: api/python/src/hse/evifluor/lookup_table.py
# SPDX-License-Identifier: MIT
# SPDX-FileCopyrightText: © 2025 HSE AG, <opensource@hseag.com>

import csv
import json
import os


class LookupTable:
    """Stores lookup table points and converts between CSV and JSON representations."""

    def __init__(self, entries = None):
        """Initializes the lookup table with optional point entries.

        Args:
            entries: Iterable of lookup-table entries containing ``concentration``
                and ``signal`` values. Defaults to an empty table.
        """
        if entries is None:
            entries = []
        self.entries = sorted(
            [self._normalize_entry(entry) for entry in entries],
            key = lambda entry: entry["signal"],
        )

    def __repr__(self):
        """Returns a debug representation of the lookup table."""
        return "LookupTable(entries={})".format(self.entries)

    def to_json(self):
        """Returns the lookup table as a JSON-serializable list."""
        return [
            {
                "concentration": entry["concentration"],
                "signal": entry["signal"],
            }
            for entry in self.entries
        ]

    def save_json(self, json_path):
        """Writes the lookup table to a JSON file.

        Args:
            json_path: Destination path for the JSON file.
        """
        with open(json_path, "w", encoding = "utf-8") as handle:
            json.dump(self.to_json(), handle, indent = 2)

    @classmethod
    def load(cls, path):
        """Loads a lookup table from a CSV or JSON file.

        Args:
            path: Source file path. The extension decides whether CSV or JSON
                loading is used.

        Returns:
            A populated :class:`LookupTable` instance.
        """
        _, extension = os.path.splitext(path)
        extension = extension.lower()

        if extension == ".json":
            return cls.load_json(path)
        if extension == ".csv":
            return cls.from_csv(path)

        raise ValueError("Unsupported lookup table file extension: {}".format(extension or "<none>"))

    @classmethod
    def load_json(cls, json_path):
        """Loads a lookup table from a JSON file.

        Args:
            json_path: Path to a JSON file containing a lookup-table list or a
                wrapper object with ``lookupTable``.

        Returns:
            A populated :class:`LookupTable` instance.
        """
        with open(json_path, "r", encoding = "utf-8") as handle:
            data = json.load(handle)

        if isinstance(data, dict):
            data = data.get("lookupTable")

        if not isinstance(data, list):
            raise ValueError("Lookup table JSON must contain a list of entries")

        return cls(data)

    @classmethod
    def from_csv(cls, csv_path):
        """Builds a lookup table from a CSV file.

        Args:
            csv_path: Path to a CSV file containing ``RFU`` and
                ``Concentration``. The first data row is interpreted as
                ``std high`` and the second data row as ``std low`` to derive
                normalized signal values.

        Returns:
            A populated :class:`LookupTable` instance.
        """
        rows = []
        with open(csv_path, "r", encoding = "utf-8", newline = "") as handle:
            reader = csv.DictReader(handle, delimiter = ";")

            if reader.fieldnames is None:
                raise ValueError("CSV file is empty")

            required_columns = {"RFU", "Concentration"}
            missing_columns = required_columns.difference(reader.fieldnames)
            if missing_columns:
                raise ValueError("CSV file is missing columns: {}".format(", ".join(sorted(missing_columns))))

            for row_index, row in enumerate(reader, start = 2):
                concentration_text = (row.get("Concentration") or "").strip()
                rfu_text = (row.get("RFU") or "").strip()

                if rfu_text == "" and concentration_text == "":
                    continue

                if concentration_text == "" or rfu_text == "":
                    raise ValueError("CSV row {} must contain RFU and Concentration".format(row_index))

                rows.append({
                    "concentration": float(concentration_text),
                    "rfu": float(rfu_text),
                })

        if len(rows) < 2:
            raise ValueError("CSV file must contain at least two data rows to derive RFU normalization")

        rfu_std_high = rows[0]["rfu"]
        rfu_std_low = rows[1]["rfu"]
        if rfu_std_high == rfu_std_low:
            raise ValueError("RFU standard high and standard low must be different")

        entries = [
            {
                "concentration": row["concentration"],
                "signal": (row["rfu"] - rfu_std_low) / (rfu_std_high - rfu_std_low),
            }
            for row in rows
        ]

        return cls(entries)

    @classmethod
    def create_json_from_csv(cls, csv_path, json_path):
        """Creates a JSON lookup-table file from an RFU CSV file.

        Args:
            csv_path: Path to the RFU CSV file.
            json_path: Destination path for the generated JSON file.

        Returns:
            The created :class:`LookupTable` instance.
        """
        lookup_table = cls.from_csv(csv_path)
        lookup_table.save_json(json_path)
        return lookup_table

    @staticmethod
    def _normalize_entry(entry):
        """Normalizes a lookup-table entry to the expected internal schema.

        Args:
            entry: Mapping containing ``concentration`` and ``signal``.

        Returns:
            A normalized dictionary with floating-point values.
        """
        if not isinstance(entry, dict):
            raise ValueError("Lookup table entries must be objects")

        if "concentration" not in entry or "signal" not in entry:
            raise ValueError("Lookup table entries must contain concentration and signal")

        return {
            "concentration": float(entry["concentration"]),
            "signal": float(entry["signal"]),
        }
